pyirf/interpolation.py

```diff
@@ -2,22 +2,245 @@
+
+    Parameters
+    ----------
+    cdfs: numpy.ndarray, shape=(N,...,M)
+        Corresponding cdf-values for all quantiles
+
+    edges: numpy.ndarray, shape=(M+1)
+        Binning of the given binned pdf
+
+    quantiles: numpy.ndarray, shape=(L)
+        L quantiles for which the ppf_values are known
+
+    Returns
+    -------
+    ppfs: numpy.ndarray, shape=(1,...,L)
+        Corresponding ppf-values for all quantiles at the target interpolation point
+    """
+
+    def cropped_interp(cdf):
+        """
+        create ppf-values through inverse interpolation of the cdf, avoiding repeating 0 and 1 entries
+        around the first and last bins as well as repeating values due to zero bins
+        in between filled bins. Both cases would result in division by zero errors when computing
+        the interpolation polynom.
+        """
+
+        # Find last 0 and first 1 entry
+        last_0 = np.nonzero(cdf == 0)[0][-1] if cdf[0] == 0 else 0
+        first_1 = np.nonzero(cdf == 1.0)[0][0]
+
+        # Predefine selection mask
+        selection = np.ones(len(cdf), dtype=bool)
+
+        # Keep only the first of subsequently matching values
+        selection[1:] = cdf[1:] != cdf[:-1]
+
+        # Keep only the last 0 and first 1 entry
+        selection[:last_0] = False
+        selection[last_0] = True
+        selection[first_1 + 1 :] = False
+        selection[first_1] = True
+
+        # create ppf values from selected bins
+        return interp1d(
+            cdf[selection],
+            bin_mids[selection],
+            bounds_error=False,
+            fill_value="extrapolate",
+        )(quantiles)
+
+    bin_mids = bin_center(edges)
+
+    # create ppf values from cdf samples via interpolation of the cdfs
+    # return nan for empty pdfs
+    ppfs = np.apply_along_axis(
+        lambda cdf: cropped_interp(cdf)
+        if np.sum(cdf) > 0
+        else np.full_like(quantiles, np.nan),
+        -1,
+        cdfs,
+    )
+    # nD interpolation of ppf values
+    return ppfs
+
+
+def pdf_from_ppf(quantiles, ppfs, edges):
+    """
+    Reconstruct pdf from ppf and evaluate at desired points.
+
+    Parameters
+    ----------
+    quantiles: numpy.ndarray, shape=(L)
+        L quantiles for which the ppf_values are known
+
+    ppfs: numpy.ndarray, shape=(1,...,L)
+        Corresponding ppf-values for all quantiles
+
+    edges: numpy.ndarray, shape=(M+1)
+        Binning of the desired binned pdf
+
+    Returns
+    -------
+    pdf_values: numpy.ndarray, shape=(1,...,M)
+        Recomputed, binned pdf
+    """
+    # recalculate pdf values through numerical differentiation
+    pdf_interpolant = np.nan_to_num(np.diff(quantiles) / np.diff(ppfs, axis=-1))
+
+    # Unconventional solution to make this usable with np.apply_along_axis for readability
+    # The ppf bin-mids are computed since the pdf-values are derived through derivation
+    # from the ppf-values
+    xyconcat = np.concatenate(
```

Comment:
Can't we just call it twice?

Reply:
Or find another solution. I really don't like this "clever" solution, especially since the array is split again into the two parts again in the lambda